File: src/main/java/com/slavsquatsuperstar/mayonez/Logger.java
package com.slavsquatsuperstar.mayonez;

import com.slavsquatsuperstar.mayonez.assets.TextFile;

import java.io.File;
import java.time.LocalDate;
import java.util.IllegalFormatException;

public final class Logger {

    public static boolean saveLogs = true;
    static String logFilename;
    private static TextFile logFile;

    static {
        if (saveLogs) {
            // Create the log directory if needed
            File logDirectory = new File("logs/");
            if (!logDirectory.exists())
                logDirectory.mkdir();

            // Count number of log files with the same date
            LocalDate today = LocalDate.now();
            int logCount = 0;
            for (File f : logDirectory.listFiles())
                if (f.getName().startsWith(today.toString()))
                    logCount++;

            logFilename = String.format("%s/%s-%d.log", logDirectory.getPath(), today, ++logCount);
            logFile = new TextFile(logFilename, false);
        }
    }

    private Logger() {}

    public static void log(Object msg, Object... args) {
        String output = String.format("[%.4f] ", Game.getTime());
        try {
            output += String.format("%s", String.format(msg.toString(), args));
            if (saveLogs)
                logFile.append(output);
        } catch (NullPointerException e) {
            output += "null";
        } catch (IllegalFormatException e) {
            output += "Logger: Could not format message";
        } finally {
            System.out.println(output);
        }
    }

    public static void log(Object msg) {
        log(msg, new Object[0]);
    }

}

File: src/main/java/slavsquatsuperstar/mayonez/Logger.kt
package slavsquatsuperstar.mayonez

import slavsquatsuperstar.mayonez.assets.TextFile
import java.io.File
import java.time.LocalDate
import java.util.*

/**
 * Prints messages to the console that can be formatted and given a priority level.
 *
 * @author SlavSquatSuperstar
 */
object Logger {

    @JvmField
    val saveLogs = Preferences.SAVE_LOGS

    @JvmField
    var logFilename: String? = null
    private var logFile: TextFile? = null

    init {
        if (saveLogs) {
            // Create the log directory if needed
            val logsDirectory = File(Preferences.LOGS_DIRECTORY)
            if (!logsDirectory.exists())
                logsDirectory.mkdir()

            // Count number of log files with the same date
            val today = LocalDate.now()
            var logCount = 0
            for (f in logsDirectory.listFiles()!!)
                if (f.name.startsWith(today.toString()))
                    logCount++

            logFilename = "${logsDirectory.path}/$today-${++logCount}.log"
            logFile = TextFile(logFilename, false)
        }
    }

    /**
     * Prints a formatted message to the console.
     *
     * @param fmt  a formatted string
     * @param args (optional) string format arguments
     */
    @JvmStatic
    fun log(fmt: Any?, vararg args: Any?) {
        val output = StringBuilder("[%02d:%.4f] ".format((Game.getTime() / 60).toInt(), Game.getTime() % 60))
        try {
            output.append(fmt.toString().format(*args))
            if (saveLogs)
                logFile!!.append(output.toString())
        } catch (e: IllegalFormatException) {
            output.append("Logger: Could not format message \"$fmt\"")
        } finally {
            println(output.toString())
        }
    }

    /**
     * Prints a message to the console.
     *
     * @param msg an object
     */
    @JvmStatic
    fun log(msg: Any?) {
        log(msg, null)
    }

    /**
     * Prints a warning to the console.
     *
     * @param msg an object
     */
    @JvmStatic
    fun warn(msg: Any?, vararg args: Any?) {
        log("[WARNING] $msg", *args)
    }

}